src/annotator/util/buckets.js

```diff
@@ -0,0 +1,248 @@
+
+    // See if this point's anchor is already in our working set of open anchors
+    const anchorIndex = current.anchors.indexOf(anchor);
+
+    if (positionType === 'start') {
+      if (anchorIndex === -1) {
+        // Add an entry for this anchor to our current set of "open" anchors
+        current.anchors.unshift(anchor);
+        current.counts.unshift(1);
+      } else {
+        // Increment the number of times we've seen a start/top edge for this
+        // anchor
+        current.counts[anchorIndex]++;
+      }
+    } else {
+      // positionType = 'end'
+      // This is the bottom/end of an anchor that we should have already seen
+      // a top edge for. Decrement the count, representing that we've found an
+      // end point to balance a previously-seen start point
+      current.counts[anchorIndex]--;
+      if (current.counts[anchorIndex] === 0) {
+        // All start points for this anchor have been balanced by end point(s)
+        // So we can remove this anchor from our collection of open anchors
+        current.anchors.splice(anchorIndex, 1);
+        current.counts.splice(anchorIndex, 1);
+      }
+    }
+
+    // For each point, we'll either:
+    // * create a new bucket: Add a new bucket (w/corresponding bucket position)
+    //   and add the working anchors to the new bucket. This, of course, has
+    //   the effect of making the buckets collection larger. OR:
+    // * merge buckets: In most cases, merge the anchors from the last bucket
+    //   into the penultimate (previous) bucket and remove the last bucket (and
+    //   its corresponding `bucketPosition` entry). Also add the working anchors
+    //   to the previous bucket. Note that this decreases the size of the
+    //   buckets collection.
+    // The ultimate set of buckets is defined by the pattern of creating and
+    // merging/removing buckets as we iterate over points.
+    const isFirstOrLastPoint =
+      bucketPositions.length === 0 || index === points.length - 1;
+
+    const isLargeGap =
+      bucketPositions.length &&
+      position - bucketPositions[bucketPositions.length - 1] > BUCKET_GAP_SIZE;
+
+    if (current.anchors.length === 0 || isFirstOrLastPoint || isLargeGap) {
+      // Create a new bucket, because:
+      // - There are no more open/working anchors, OR
+      // - This is the first or last point, OR
+      // - There's been a large dimensional gap since the last bucket's position
+      buckets.push(current.anchors.slice());
+      // Each bucket gets a corresponding entry in `bucketPositions` for the
+      // pixel position of its eventual indicator in the bucket bar
+      bucketPositions.push(position);
+    } else {
+      // Merge buckets
+      // We will remove 2 (usually) or 1 (if there is only one) bucket
+      // from the buckets collection, and re-add 1 merged bucket (always)
+
+      // Always pop off the last bucket
+      const ultimateBucket = buckets.pop() || new Array();
```

Comment:
`|| new Array()` here and below are a TS affordance because of strict null even though there's no way to get to this point without `buckets` having at least one Array of `Anchor`s in it.